index.js

```diff
@@ -1,241 +1,232 @@
+        env.IS_APP_ENGINE = !!(process.env.GAE_SERVICE || process.env.GAE_MODULE_NAME);
+      }
 
-    client.getAccessToken(callback);
-  });
-};
+      callback(null, env.IS_APP_ENGINE);
+    });
+  }
 
-Auth.prototype.isAppEngine = function (callback) {
-  var self = this;
+  isCloudFunction (callback) {
+    setImmediate(() => {
+      var env = this.environment;
 
-  setImmediate(function () {
-    if (typeof self.environment.IS_APP_ENGINE === 'undefined') {
-      self.environment.IS_APP_ENGINE =
-        !!(process.env.GAE_SERVICE || process.env.GAE_MODULE_NAME);
-    }
+      if (typeof env.IS_CLOUD_FUNCTION === 'undefined') {
+        env.IS_CLOUD_FUNCTION = !!process.env.FUNCTION_NAME;
+      }
 
-    callback(null, self.environment.IS_APP_ENGINE);
-  });
-};
+      callback(null, env.IS_CLOUD_FUNCTION);
+    });
+  }
 
-Auth.prototype.isCloudFunction = function (callback) {
-  var self = this;
+  isComputeEngine (callback) {
+    var env = this.environment;
 
-  setImmediate(function () {
-    if (typeof self.environment.IS_CLOUD_FUNCTION === 'undefined') {
-      self.environment.IS_CLOUD_FUNCTION = !!process.env.FUNCTION_NAME;
+    if (typeof env.IS_COMPUTE_ENGINE !== 'undefined') {
+      setImmediate(() => {
+        callback(null, env.IS_COMPUTE_ENGINE);
+      });
+      return;
     }
 
-    callback(null, self.environment.IS_CLOUD_FUNCTION);
-  });
-};
-
-Auth.prototype.isComputeEngine = function (callback) {
-  var self = this;
+    request('http://metadata.google.internal', (err, res) => {
+      env.IS_COMPUTE_ENGINE = !err && res.headers['metadata-flavor'] === 'Google';
 
-  if (typeof this.environment.IS_COMPUTE_ENGINE !== 'undefined') {
-    setImmediate(function () {
-      callback(null, self.environment.IS_COMPUTE_ENGINE);
+      callback(null, env.IS_COMPUTE_ENGINE);
     });
-    return;
   }
 
-  request('http://metadata.google.internal', function (err, res) {
-    self.environment.IS_COMPUTE_ENGINE =
-      !err && res.headers['metadata-flavor'] === 'Google';
+  isContainerEngine (callback) {
+    var env = this.environment;
 
-    callback(null, self.environment.IS_COMPUTE_ENGINE);
-  });
-};
+    if (typeof env.IS_CONTAINER_ENGINE !== 'undefined') {
+      setImmediate(() => {
+        callback(null, env.IS_CONTAINER_ENGINE);
+      });
+      return;
+    }
 
-Auth.prototype.isContainerEngine = function (callback) {
-  var self = this;
+    gcpMetadata.instance('/attributes/cluster-name', err => {
+      env.IS_CONTAINER_ENGINE = !err;
 
-  if (typeof this.environment.IS_CONTAINER_ENGINE !== 'undefined') {
-    setImmediate(function () {
-      callback(null, self.environment.IS_CONTAINER_ENGINE);
+      callback(null, env.IS_CONTAINER_ENGINE);
     });
-    return;
   }
+}
 
-  gcpMetadata.instance('/attributes/cluster-name', function (err) {
-    self.environment.IS_CONTAINER_ENGINE = !err;
-
-    callback(null, self.environment.IS_CONTAINER_ENGINE);
-  });
+module.exports = config => {
```

Comment:
I'm not very experienced with some of the new ES features (like Classes). Is there a way to support lazy instantiation, e.g.:

```js
var googleAutoAuth = require('google-auto-auth')

var authClient = new googleAutoAuth()
// and
var authClient = googleAutoAuth()
```